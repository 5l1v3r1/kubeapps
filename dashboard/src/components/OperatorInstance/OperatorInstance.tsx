import { RouterAction } from "connected-react-router";
import * as yaml from "js-yaml";
import * as React from "react";

import placeholder from "../../placeholder.png";
import { IClusterServiceVersion, IClusterServiceVersionCRD, IResource } from "../../shared/types";
import AppNotes from "../AppView/AppNotes";
import AppValues from "../AppView/AppValues";
import Card, { CardContent, CardFooter, CardGrid, CardIcon } from "../Card";
import ConfirmDialog from "../ConfirmDialog";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";

export interface IOperatorInstanceProps {
  isFetching: boolean;
  namespace: string;
  csvName: string;
  crdName: string;
  instanceName: string;
  getResource: (
    namespace: string,
    csvName: string,
    crdName: string,
    resourceName: string,
  ) => Promise<void>;
  deleteResource: (namespace: string, crdName: string, resource: IResource) => Promise<boolean>;
  push: (location: string) => RouterAction;
  error?: Error;
  resource?: IResource;
  csv?: IClusterServiceVersion;
}

export interface IOperatorInstanceState {
  modalIsOpen: boolean;
  crd?: IClusterServiceVersionCRD;
}

class OperatorInstance extends React.Component<IOperatorInstanceProps, IOperatorInstanceState> {
  public state: IOperatorInstanceState = {
    modalIsOpen: false,
  };

  public componentDidMount() {
    const { csvName, crdName, instanceName, namespace, getResource } = this.props;
    getResource(namespace, csvName, crdName, instanceName);
  }

  public componentDidUpdate(prevProps: IOperatorInstanceProps) {
    const { csvName, crdName, instanceName, namespace, getResource, resource, csv } = this.props;
    if (prevProps.namespace !== namespace) {
      getResource(namespace, csvName, crdName, instanceName);
      return;
    }

    if (csv !== prevProps.csv || resource !== prevProps.resource) {
      if (csv && resource) {
        this.setState({
          crd: csv.spec.customresourcedefinitions.owned.find(c => c.kind === resource.kind),
        });
      }
    }
  }

  public render() {
    const { isFetching, error, resource, csv, instanceName, namespace } = this.props;
    return (
      <section className="AppView padding-b-big">
        <main>
          <LoadingWrapper loaded={!isFetching}>
            {error && (
              <ErrorSelector
                error={error}
                action="get"
                resource={`Operator intance ${instanceName}`}
                namespace={namespace}
              />
            )}
            {resource && (
              <div className="row collapse-b-tablet">
                <div className="col-3">{csv && this.renderCSVInfo(csv)}</div>
                <div className="col-9">
                  <div className="row padding-t-bigger">
                    <div className="col-4">
                      <h5>{instanceName}</h5>
                    </div>
                    <div className="col-8 text-r">
                      <button className="button button-danger" onClick={this.openModal}>
                        Delete
                      </button>
                      <ConfirmDialog
                        onConfirm={this.handleDeleteClick}
                        modalIsOpen={this.state.modalIsOpen}
                        loading={this.props.isFetching}
                        closeModal={this.closeModal}
                      />
                    </div>
                  </div>
                  <AppNotes title="Status" notes={yaml.safeDump(resource.status)} />
                  <AppValues values={yaml.safeDump(resource.spec)} />
                </div>
              </div>
            )}
          </LoadingWrapper>
        </main>
      </section>
    );
  }

  private renderCSVInfo = (csv: IClusterServiceVersion) => {
    const { instanceName } = this.props;
    const { crd } = this.state;
    const icon = csv.spec.icon?.length
      ? `data:${csv.spec.icon[0].mediatype};base64,${csv.spec.icon[0].base64data}`
      : placeholder;
    return (
      <CardGrid className="ChartInfo">
        <Card>
          <CardIcon icon={icon} />
          <CardContent>
            <h5>{instanceName}</h5>
            <p className="margin-b-reset">{crd?.description}</p>
          </CardContent>
          <CardFooter>
            <div>
              <div>Cluster Service Version: v{csv.spec.version}</div>
              <div>Kind: {crd?.kind}</div>
            </div>
          </CardFooter>
        </Card>
      </CardGrid>
    );
  };

  private openModal = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  private closeModal = async () => {
    this.setState({
      modalIsOpen: false,
    });
  };

  private handleDeleteClick = async () => {
    const { namespace, resource } = this.props;
    const { crd } = this.state;
    const deleted = await this.props.deleteResource(namespace, crd!.name.split(".")[0], resource!);
    this.closeModal();
    if (deleted) {
      this.props.push(`/apps/ns/${namespace}`);
    }
  };
}

export default OperatorInstance;
