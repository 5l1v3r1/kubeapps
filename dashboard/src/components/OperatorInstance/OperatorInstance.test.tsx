import { shallow } from "enzyme";
import * as React from "react";
import * as ReactModal from "react-modal";
import itBehavesLike from "../../shared/specs";
import AppNotes from "../AppView/AppNotes";
import AppValues from "../AppView/AppValues";
import ConfirmDialog from "../ConfirmDialog";
import { ErrorSelector } from "../ErrorAlert";
import OperatorInstance, { IOperatorInstanceProps } from "./OperatorInstance";

const defaultProps: IOperatorInstanceProps = {
  isFetching: false,
  namespace: "default",
  csvName: "foo",
  crdName: "foo.kubeapps.com",
  instanceName: "foo-cluster",
  getResource: jest.fn(),
  deleteResource: jest.fn(),
  push: jest.fn(),
};

itBehavesLike("aLoadingComponent", {
  component: OperatorInstance,
  props: { ...defaultProps, isFetching: true },
});

it("gets a resource when loading the component", () => {
  const getResource = jest.fn();
  shallow(<OperatorInstance {...defaultProps} getResource={getResource} />);
  expect(getResource).toHaveBeenCalledWith(
    defaultProps.namespace,
    defaultProps.csvName,
    defaultProps.crdName,
    defaultProps.instanceName,
  );
});

it("gets a resource again if the namespace changes", () => {
  const getResource = jest.fn();
  const wrapper = shallow(<OperatorInstance {...defaultProps} getResource={getResource} />);
  wrapper.setProps({ namespace: "other" });
  expect(getResource).toHaveBeenCalledTimes(2);
});

it("renders an error", () => {
  const wrapper = shallow(<OperatorInstance {...defaultProps} error={new Error("Boom!")} />);
  expect(wrapper.find(ErrorSelector)).toExist();
  expect(wrapper.find(AppNotes)).not.toExist();
});

describe("renders a resource", () => {
  const csv = {
    metadata: { name: "foo" },
    spec: {
      customresourcedefinitions: {
        owned: [{ name: "foo.kubeapps.com", version: "v1alpha1", kind: "Foo" }],
      },
    },
  } as any;
  const resource = { kind: "Foo", spec: { test: true }, status: { alive: true } } as any;
  const props = { ...defaultProps, csv, resource };

  it("renders the resource and CSV info", () => {
    const wrapper = shallow(<OperatorInstance {...props} />);
    expect(wrapper.find(AppNotes)).toExist();
    expect(wrapper.find(AppValues)).toExist();
    expect(wrapper.find(".ChartInfo")).toExist();
    expect(wrapper).toMatchSnapshot();
  });

  it("deletes the resource", async () => {
    const deleteResource = jest.fn(() => true);
    const push = jest.fn();
    const wrapper = shallow(
      <OperatorInstance {...props} deleteResource={deleteResource} push={push} />,
    );
    wrapper.setState({ crd: csv.spec.customresourcedefinitions.owned[0] });
    ReactModal.setAppElement(document.createElement("div"));
    wrapper.find(".button-danger").simulate("click");

    const dialog = wrapper.find(ConfirmDialog);
    expect(dialog.prop("modalIsOpen")).toEqual(true);
    (dialog.prop("onConfirm") as any)();
    expect(deleteResource).toHaveBeenCalledWith(defaultProps.namespace, "foo", resource);
    // wait async calls
    await new Promise(r => r());
    expect(push).toHaveBeenCalledWith(`/apps/ns/${defaultProps.namespace}`);
  });
});
