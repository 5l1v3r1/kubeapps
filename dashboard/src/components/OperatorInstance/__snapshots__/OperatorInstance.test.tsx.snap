// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<section
  className="AppView padding-b-big"
>
  <main>
    <LoadingWrapper
      loaded={false}
      type={0}
    />
  </main>
</section>
`;

exports[`renders a resource renders the resource and CSV info 1`] = `
<section
  className="AppView padding-b-big"
>
  <main>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <div
        className="row collapse-b-tablet"
      >
        <div
          className="col-3"
        >
          <CardGrid
            className="ChartInfo"
          >
            <Card>
              <CardIcon />
              <CardContent>
                <h5>
                  foo-cluster
                </h5>
                <p
                  className="margin-b-reset"
                />
              </CardContent>
              <CardFooter>
                <div>
                  <div>
                    Cluster Service Version: v
                  </div>
                  <div>
                    Kind: 
                  </div>
                </div>
              </CardFooter>
            </Card>
          </CardGrid>
        </div>
        <div
          className="col-9"
        >
          <div
            className="row padding-t-bigger"
          >
            <div
              className="col-4"
            >
              <h5>
                foo-cluster
              </h5>
            </div>
            <div
              className="col-8 text-r"
            >
              <button
                className="button button-danger"
                onClick={[Function]}
              >
                Delete
              </button>
              <ConfirmDialog
                closeModal={[Function]}
                loading={false}
                modalIsOpen={false}
                onConfirm={[Function]}
              />
            </div>
          </div>
          <AppNotes
            notes="alive: true
          "
            title="Status"
          />
          <AppValues
            values="test: true
          "
          />
        </div>
      </div>
    </LoadingWrapper>
  </main>
</section>
`;
